### PR TITLE
add interface to stepdown or join leader election

### LINKED ISSUE #2422: Provide a way to disallow a node from being stream processor leader
If a log stream has be restored from a remote node, and the restore operation uses a snapshot, then until the restore operation is finished the node cannot be stream processor leader for this log stream, since the snapshot is not valid until the necessary events for the snapshot to be valid have been also been replicated.

AT:

- There is a way to block a node from being stream processor leader for a given partition, and to later allow it to be

### PR DESCRIPTION
**Pull request by deepthidevaki**
`PartitionLeaderElection` provides an interface to stepdown and join the election on demand. This can be used by other modules,  for example #2424 can decide to step down if the node trying to restore is itself the leader.

- Also did some refactoring to address the issue in #2414 
closes #2422 #2414

### CODE REVIEW

File: broker-core/src/main/java/io/zeebe/broker/clustering/base/partitions/PartitionLeaderElection.java [hunk 1]
```diff
@@ -81,35 +82,75 @@ public void start(ServiceStartContext startContext) {
...
+  }
 
-    startContext.async(startFuture, true);
+  private void tryJoinElection(CompletableActorFuture<Void> joinFuture) {
+    try {
+      if (canBecomeLeader) {
+        election.run(memberId);
+      }
+      joinFuture.complete(null);
+      actor.done();
```
Comment (npepinpe) [inline broker-core/src/main/java/io/zeebe/broker/clustering/base/partitions/PartitionLeaderElection.java]: It's a little confusing for me that it calls `actor.done` and `actor.yield` outside of a `actor.runUntilDone`. How about this method returns a true/false if it joined, and wherever you call it in a `runUntilDone` you can check the result and decide whether to yield or not? Or a simpler solution, if these methods (`tryAddListener` and `tryJoinElection` are only used in `runUntilDone`, why not move the `runUntilDone` call in them? That way it's clearer that you **have** to call `done` or `yield`.

File: broker-core/src/main/java/io/zeebe/broker/clustering/base/partitions/PartitionLeaderElection.java [hunk 2]
```diff
@@ -134,6 +175,39 @@ private void transitionToLeader(long term) {
...
+   */
+  public CompletableActorFuture<Void> stepdown() {
+    final CompletableActorFuture<Void> future = new CompletableActorFuture();
+    actor.call(
+        () -> {
+          canBecomeLeader = false;
+          try {
+            election.withdraw(memberId);
+            future.complete(null);
+          } catch (PrimitiveException e) {
```
Comment (npepinpe) [inline broker-core/src/main/java/io/zeebe/broker/clustering/base/partitions/PartitionLeaderElection.java]: I guess we'd want to complete exceptionally if any exception is thrown at all?

**Review by npepinpe — COMMENTED** [review]
Looks good, like I mentioned I think `actor.call` is used to return a result, and you can probably just use `actor.run` or `actor.submit` in most places where you use call. Try it, I'm only 90% sure, I'm not that familiar with the scheduler.

**Review by npepinpe — APPROVED** [review]

**Comment by deepthidevaki** [comment]
bors retry

**Pull request merged**